inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -0,0 +1,206 @@
+    }
+}  // namespace
+
+AutoInferencePlugin::AutoInferencePlugin() {
+    _pluginName = "AUTO";
+}
+
+IE::ExecutableNetworkInternal::Ptr AutoInferencePlugin::LoadExeNetworkImpl(const IE::CNNNetwork& network, const ConfigType& config) {
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto metaDevices = GetDeviceChoice(fullConfig);
+
+    // FIXME: always select CPU device now
+    DeviceInformation selectedDevice = SelectDevice(metaDevices);
+    IE::ExecutableNetwork executableNetwork;
+    try {
+        executableNetwork = GetCore()->LoadNetwork(network, selectedDevice.deviceName, selectedDevice.config);
+    } catch(const IE::Exception &iie) {
+        IE_THROW() << "Failed to load network to device named " << selectedDevice.deviceName;
+    }
+
+    bool enablePerfCounters = false;
+    try {
+        enablePerfCounters =
+            executableNetwork.GetConfig(IE::PluginConfigParams::KEY_PERF_COUNT).as<std::string>() ==
+                IE::PluginConfigParams::YES;
+    } catch (...) {
+    }
+
+    return std::make_shared<AutoExecutableNetwork>(executableNetwork,
+                                                   selectedDevice,
+                                                   enablePerfCounters);
+}
+
+IE::QueryNetworkResult AutoInferencePlugin::QueryNetwork(const IE::CNNNetwork& network, const ConfigType& config) const {
+    IE::QueryNetworkResult queryResult;
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    queryResult.rc = IE::StatusCode::OK;
+    queryResult.supportedLayersMap.clear();
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto metaDevices = GetDeviceChoice(fullConfig);
+    std::unordered_set<std::string> supportedLayers;
+    std::unordered_set<std::string> supportedDevices;
+    for (auto&& value : metaDevices) {
+        try {
+            if (value.deviceName != _pluginName) {
+                auto deviceQr = GetCore()->QueryNetwork(network, value.deviceName, value.config);
+                std::unordered_set<std::string> deviceSupportedLayers;
+                for (auto &&layerQr : deviceQr.supportedLayersMap) {
+                    deviceSupportedLayers.emplace(layerQr.first);
+                }
+                supportedLayers = supportedLayers.empty()
+                                ? deviceSupportedLayers : (deviceSupportedLayers.empty()
+                                ? supportedLayers : IE::details::Intersection(
+                                     supportedLayers, deviceSupportedLayers));
```

Comment:
why AUTO supports all layers from device plugins? It uses only a single plugin to load network...

Reply:
since we don't know which device to load network during query, so we suppose query all available devices. should we add selete logic to query?

Reply:
But what if plugin A supports only operation a, plugin B supports operation b.
But network consists of a -> b. How to support such network? Are we going to use HETERO plugin for such case? 

I think we need to select device first and call QueryNetwork for such device plugin. The results of such query network is returned as QueryNetwork result for AUTO itself.

Reply:
Agree with Ilya that AUTO need query whole network on each individual device plugin. if one candidate device can support this network, the result of this candidate device is updated and returned as AUTO result.

Reply:
already modified in #5366